host/ecallids.c

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "ecallids.h"
+#include <openenclave/internal/raise.h>
+#include "hostthread.h"
+
+static const char* _ecall_table[OE_MAX_ECALLS];
+static uint32_t _ecall_table_size;
+static oe_mutex _lock;
+
+// Get the global ecall id from the _ecall_table.
+static oe_result_t _get_global_ecall_id_by_name(
+    const char* name,
+    uint64_t* global_id)
+{
+    oe_result_t result = OE_NOT_FOUND;
+    bool locked = false;
+    uint64_t i;
+
+    if (oe_mutex_lock(&_lock) != 0)
+        goto done;
+    locked = true;
+
+    if (!global_id)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    for (i = 0; i < _ecall_table_size; i++)
+    {
+        if (strcmp(_ecall_table[i], name) == 0)
+        {
+            *global_id = i;
+            result = OE_OK;
+            break;
+        }
+    }
+
+    /* If the name is not found, adding it to the table. */
+    if (result == OE_NOT_FOUND)
+    {
+        if (_ecall_table_size >= OE_MAX_ECALLS)
+        {
+            result = OE_OUT_OF_BOUNDS;
+            goto done;
+        }
+
+        _ecall_table[_ecall_table_size] = name;
+        *global_id = _ecall_table_size;
+        _ecall_table_size++;
+        result = OE_OK;
+    }
+
+done:
+    if (locked)
+        oe_mutex_unlock(&_lock);
+    return result;
+}
+
+oe_result_t oe_get_enclave_function_id(
+    oe_enclave_t* enclave,
+    uint64_t* global_id,
+    const char* name,
+    uint64_t* id)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_ecall_id_t* ecall_id_table = NULL;
+
+    if (!enclave || !global_id || !name || !id)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        if ((result = _get_global_ecall_id_by_name(name, global_id)) != OE_OK)
+            goto done;
+    }
+
+    if (*global_id >= OE_MAX_ECALLS)
+    {
+        result = OE_OUT_OF_BOUNDS;
+        goto done;
+    }
+
+    OE_CHECK(oe_get_ecall_id_table(enclave, &ecall_id_table));
```

Comment:
Consider adding comments here explaining why we need to do this (i.e.., we cannot directly obtain via `enclave->ecall_id_table` because of the `oe_enclave_t` is opaque here).

Reply:
Done.